src/parallel/mpi_concurrency/mpi_processor_grouping.cpp

```diff
@@ -0,0 +1,135 @@
+// Copyright (C) 2018 ETH Zurich
+// Copyright (C) 2018 UT-Battelle, LLC
+// All rights reserved.
+//
+// See LICENSE.txt for terms of usage.
+// See CITATION.txt for citation guidelines if you use this code for scientific publications.
+//
+// Author: Giovanni Balduzzi (gbalduzz@itp.phys.ethz.ch)
+//
+// This file implements mpi_processor_grouping.hpp.
+
+#include "dca/parallel/mpi_concurrency/mpi_processor_grouping.hpp"
+
+#include <iostream>
+#include <unistd.h>
+
+#include "dca/parallel/mpi_concurrency/kernel_test.hpp"
+
+namespace dca {
+namespace parallel {
+// dca::parallel::
+
+MPIProcessorGrouping::MPIProcessorGrouping() : MPIProcessorGrouping(defaultTest) {}
+
+MPIProcessorGrouping::MPIProcessorGrouping(bool (*test)()) {
+  // Initialize grouping with MPI world.
+  MPI_communication_ = MPI_COMM_WORLD;
+  MPI_Comm_size(MPI_COMM_WORLD, &world_size_);
+  MPI_Comm_rank(MPI_COMM_WORLD, &world_id_);
+
+  // Check if the process has the desired qualities, or remove it from the communicator.
+  const bool is_valid = test();
+
+  std::vector<char> validity_input(world_size_, 0);
+  std::vector<char> validity_output(world_size_, 0);
+  validity_input[world_id_] = is_valid;
+  MPI_Allreduce(validity_input.data(), validity_output.data(), validity_input.size(), MPI_CHAR,
+                MPI_MAX, MPI_COMM_WORLD);
+
+  std::vector<int> valid_ids;
+  for (int id = 0; id < validity_output.size(); ++id)
+    if (validity_output[id])
+      valid_ids.push_back(id);
+  if (!valid_ids.size())
+    throw(std::logic_error("MPI grouping is empty.\n"));
+
+  MPI_Group world_group;
+  MPI_Comm_group(MPI_COMM_WORLD, &world_group);
+  MPI_Group_incl(world_group, valid_ids.size(), valid_ids.data(), &MPI_group_);
+  MPI_Comm_create_group(MPI_COMM_WORLD, MPI_group_, 1, &MPI_communication_);
+  MPI_Group_free(&world_group);
+
+  if (is_valid) {
+    MPI_Comm_size(MPI_communication_, &size_);
+    MPI_Comm_rank(MPI_communication_, &id_);
+  }
+  else {
+    id_ = size_ = -1;
+  }
+
+  if (world_size_ > size_)
+    printRemovedProcesses(valid_ids);
+}
+
+MPIProcessorGrouping::~MPIProcessorGrouping() {
+  if (isValid()) {
+    MPI_Comm_free(&MPI_communication_);
+    MPI_Group_free(&MPI_group_);
+  }
+}
+
+bool MPIProcessorGrouping::defaultTest() {
+#ifdef DCA_HAVE_CUDA
+  try {
+    return kernelTest();
```

Comment:
This is not "testing code" but rather the default test to see if a GPU is working.  Conceivably you could pass your own test kernel to the constructor. perhaps the name should be defaultCheck?  The code is underdocumented but one can read through it with blame included on the dev repo.

Since @gbalduzz its this or read.

Reply:
Yes, you can pass your own kernel and this is done mainly for testing. renaming test->check sounds good to me.